sycl/doc/design/DeviceIf.md

```diff
@@ -0,0 +1,277 @@
+does not identify a specific device name.  As a reminder, the current command
+line for AOT compilation on GPU looks like this:
+
+```
+$ clang++ -fsycl -fsycl-targets=spir64_gen -Xs "-device skl" ...
+```
+
+Notice that `-fsycl-targets` option specifies only the generic name
+`spir64_gen` whereas the device name is passed directly to `ocloc` via
+`-Xs "-device skl"`.  Since the compiler driver merely passes the `-Xs` options
+directly to the underlying `ocloc` without understanding them, the compiler
+driver does not currently know the target device(s) of the AOT compilation.
+
+To fix this, the `-fsycl-targets` option should be changed to accept the
+following GPU device names in addition to the target names it currently
+recognizes:
+
+* `intel_gpu_bdw`
+* `intel_gpu_skl`
+* `intel_gpu_kbl`
+* `intel_gpu_cfl`
+* `intel_gpu_apl`
+* `intel_gpu_glk`
+* `intel_gpu_whl`
+* `intel_gpu_aml`
+* `intel_gpu_cml`
+* `intel_gpu_icllp`
+* `intel_gpu_ehl`
+* `intel_gpu_tgllp`
+* `intel_gpu_rkl`
+* `intel_gpu_adl_s`
+* `intel_gpu_rpl_s`
+* `intel_gpu_adl_p`
+* `intel_gpu_adl_n`
+* `intel_gpu_dg1`
+* `intel_gpu_acm_g10`
+* `intel_gpu_acm_g11`
+* `intel_gpu_acm_g12`
+* `intel_gpu_pvc`
+* `intel_gpu_8_0_0` (alias for `intel_gpu_bdw`)
+* `intel_gpu_9_0_9` (alias for `intel_gpu_skl`)
+* `intel_gpu_9_1_9` (alias for `intel_gpu_kbl`)
+* `intel_gpu_9_2_9` (alias for `intel_gpu_cfl`)
+* `intel_gpu_9_3_0` (alias for `intel_gpu_apl`)
+* `intel_gpu_9_4_0` (alias for `intel_gpu_glk`)
+* `intel_gpu_9_5_0` (alias for `intel_gpu_whl`)
+* `intel_gpu_9_6_0` (alias for `intel_gpu_aml`)
+* `intel_gpu_9_7_0` (alias for `intel_gpu_cml`)
+* `intel_gpu_11_0_0` (alias for `intel_gpu_icllp`)
+* `intel_gpu_11_2_0` (alias for `intel_gpu_ehl`)
+* `intel_gpu_12_0_0` (alias for `intel_gpu_tgllp`)
+* `intel_gpu_12_10_0` (alias for `intel_gpu_dg1`)
+
+The above listed device names may not be mixed with the existing target name
+`spir64_gen` on the same command line.  In addition, the user must not pass the
+`-device` option to `ocloc` via `-Xs` or related command line options because
+the compiler driver will pass this option to `ocloc` automatically.
+
+Note that in the first phase of implementation, only one of the above listed
+GPU device names may appear on the command line.  As a result, the first phase
+of implementation supports AOT compilation in this new mode only for a single
+GPU device.
+
+
+## Phase 1: AOT mode only
+
+The first phase requires changes only to the compiler driver and to the
+device headers.
+
+### Compiler driver macro predefines
+
+Most of the changes to the compiler driver are described above, but there are
+a few small additional changes that are specific to phase 1.  If the user
+invokes the compiler driver with `-fsycl-targets=<device>` where `<device>` is
+one of the GPU device names listed above, the compiler driver must predefine
+one of the following corresponding C++ macro names:
```

Comment:
Is there a specific reason why ptx64 is skipped in aspects and macros? Would be good to specify in the doc then.

Reply:
`intel_gpu_xe` is not a specific device too, it is a family of devices with different capabilities. E.g. Both TGLLP and PVC are XE, but the first one does not support FP64. I think supporting such device families as target architecture is useful.

Reply:
But PTX and SPIR-V are implementation details.  Let's suppose (hypothetically) that we decide to change the implementation of DPC++ to stop using SPIR-V and we use some other IL instead.  Would we change the implementation of `if_device_has<spir64>()` to return `false` for every device?  Why would an application want to  run code conditionally based on an implementation detail like the IL format?

Reply:
Well, the names may coincide with IL format, but the they mean more than that. PTX actually means generic NV target, and there is no other way to specify compilation for NV device. SPIRV - generic "portable" target.

Reply:
> if_device_has<spir64>()

spirv64 is kind of special and can be skipped, as this is not really and AOT and does not mean invoking and AOT compiler, unlike the rest of targets

Reply:
I think we might want to add some way to select code based on the CUDA architecture, and probably also on the CUDA compute capability.  I think we should wait, though, and add this later.  It's not clear to me yet what we can support.  There's still a lot of refinement to do on this extension.  For example, there's no support yet for FPGA, and the CPU targets need to be expanded.  I don't think we need to add CUDA support in this first step.

Reply:
Done in 372c480989a42234eead02b6323d991e10b8b14a